File: lessons/dsa/hash-lookup-as-a-trade/examples/LookupCostsTest.java
import java.util.ArrayList;
import java.util.List;

public final class LookupCostsTest {
  private static int failures = 0;
  private static final List<String> KEYS =
      List.of("cat", "dog", "emu", "fox", "owl", "bat", "ant", "cow");

  private static void check(String name, boolean passed) {
    if (!passed) {
      System.out.println("FAIL " + name);
      failures++;
    }
  }

  public static void main(String[] args) {
    check("scan reaches the wanted key", LookupCosts.keysExaminedByScan(KEYS, "owl") == 5);
    check("scan examines all when absent",
        LookupCosts.keysExaminedByScan(KEYS, "yak") == KEYS.size());
    check("hash examines only its slot", LookupCosts.keysExaminedByHash(KEYS, "owl", 16) < 3);
    check("absent key examines only its slot",
        LookupCosts.keysExaminedByHash(KEYS, "yak", 16) < 3);

    List<String> larger = new ArrayList<>(KEYS);
    for (int index = 0; index < 200; index++) {
      larger.add("key" + index);
    }
    check("hash stays cheap as the collection grows",
        LookupCosts.keysExaminedByHash(larger, "owl", 512) < 3);
    check("scan does not", LookupCosts.keysExaminedByScan(larger, "owl") > 3);

    check("a slot is stable", LookupCosts.hashSlot("cat", 6) == LookupCosts.hashSlot("cat", 6));
    for (String key : KEYS) {
      check("a slot is inside the table",
          LookupCosts.hashSlot(key, 6) >= 0 && LookupCosts.hashSlot(key, 6) < 6);
    }

    if (failures > 0) {
      System.exit(1);
    }
    System.out.println("All checks passed.");
  }
}

File: lessons/dsa/hash-lookup-as-a-trade/examples/lookup_costs_test.cpp
#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>

#include "lookup_costs.cpp"

namespace {
int failures = 0;
const std::vector<std::string> kKeys = {"cat", "dog", "emu", "fox",
                                        "owl", "bat", "ant", "cow"};

void check(const char* name, bool passed) {
  if (!passed) {
    std::cout << "FAIL " << name << "\n";
    ++failures;
  }
}
}  // namespace

int main() {
  check("scan reaches the wanted key", keysExaminedByScan(kKeys, "owl") == 5);
  check("scan examines all when absent",
        keysExaminedByScan(kKeys, "yak") == static_cast<int>(kKeys.size()));
  check("hash examines only its slot", keysExaminedByHash(kKeys, "owl", 16) < 3);
  check("absent key examines only its slot", keysExaminedByHash(kKeys, "yak", 16) < 3);

  std::vector<std::string> larger = kKeys;
  for (int index = 0; index < 200; ++index) {
    larger.push_back("key" + std::to_string(index));
  }
  check("hash stays cheap as the collection grows", keysExaminedByHash(larger, "owl", 512) < 3);
  check("scan does not", keysExaminedByScan(larger, "owl") > 3);

  check("a slot is stable", hashSlot("cat", 6) == hashSlot("cat", 6));
  for (const std::string& key : kKeys) {
    check("a slot is inside the table", hashSlot(key, 6) >= 0 && hashSlot(key, 6) < 6);
  }

  if (failures > 0) {
    return EXIT_FAILURE;
  }
  std::cout << "All checks passed.\n";
  return EXIT_SUCCESS;
}
